core/server/master/src/main/java/alluxio/master/metastore/CachingInodeStore.java

```diff
@@ -0,0 +1,143 @@
+
+  // Cache recently-accessed inode tree edges.
+  private final Cache<Edge, Long> mEdgeCache = CacheBuilder.newBuilder()
+      .maximumSize(Configuration.getInt(PropertyKey.MASTER_METASTORE_INODE_CACHE_SIZE)).build();
+
+  /**
+   * @param backingStore the backing inode store
+   */
+  public CachingInodeStore(InodeStore backingStore) {
+    mBackingStore = backingStore;
+  }
+
+  @Override
+  public Optional<MutableInode<?>> getMutable(long id) {
+    Optional<MutableInode<?>> inode = Optional.ofNullable(mInodeCache.getIfPresent(id));
+    if (inode.isPresent()) {
+      return inode;
+    }
+    inode = mBackingStore.getMutable(id);
+    if (inode.isPresent()) {
+      mInodeCache.put(id, inode.get());
+    }
+    return inode;
+  }
+
+  @Override
+  public void remove(InodeView inode) {
+    mInodeCache.invalidate(inode.getId());
+    mEdgeCache.invalidate(new Edge(inode.getParentId(), inode.getName()));
+    mBackingStore.remove(inode);
+  }
+
+  @Override
+  public void writeInode(MutableInode<?> inode) {
+    mInodeCache.put(inode.getId(), inode);
+    mBackingStore.writeInode(inode);
+  }
+
+  @Override
+  public void clear() {
+    mInodeCache.invalidateAll();
+    mEdgeCache.invalidateAll();
+    mBackingStore.clear();
+  }
+
+  @Override
+  public void addChild(long parentId, InodeView inode) {
+    mEdgeCache.put(new Edge(parentId, inode.getName()), inode.getId());
+    mBackingStore.addChild(parentId, inode);
+  }
+
+  @Override
+  public void removeChild(long parentId, String name) {
+    mEdgeCache.invalidate(new Edge(parentId, name));
+    mBackingStore.removeChild(parentId, name);
+  }
+
+  @Override
+  public long estimateSize() {
+    return mBackingStore.estimateSize();
+  }
+
+  @Override
+  public Iterable<Long> getChildIds(InodeDirectoryView inode) {
+    return mBackingStore.getChildIds(inode);
+  }
+
+  @Override
+  public Iterable<? extends Inode> getChildren(InodeDirectoryView inode) {
+    Iterator<Long> baseIterator = mBackingStore.getChildIds(inode).iterator();
+    return () -> new Iterator<Inode>() {
+      @Override
+      public boolean hasNext() {
+        return baseIterator.hasNext();
+      }
+
+      @Override
+      public Inode next() {
+        return get(baseIterator.next()).get();
+      }
+    };
+  }
+
+  @Override
+  public Optional<Long> getChildId(InodeDirectoryView inode, String name) {
+    return mBackingStore.getChildId(inode, name);
+  }
+
+  @Override
+  public Optional<Inode> getChild(InodeDirectoryView inode, String name) {
+    Edge edge = new Edge(inode.getId(), name);
+    Long childId = mEdgeCache.getIfPresent(edge);
+    if (childId != null) {
+      return get(childId);
+    }
+    return mBackingStore.getChildId(inode, name).flatMap(id -> {
+      mEdgeCache.put(edge, id);
+      return get(id);
+    });
+  }
+
+  @Override
+  public boolean hasChildren(InodeDirectoryView inode) {
+    return mBackingStore.hasChildren(inode);
```

Comment:
We can't try for a cache hit here?

Reply:
This would require indexing the cached edges by inode id. A more efficient, but more complicated approach is to cache full directory listings (indexed by parent id), which will allow us to implement both `hasChildren` and `getChildren` using only the cache.